### PR TITLE
Custom solver

### PR DESCRIPTION
**Pull request by KulaginVladimir**
## Proposed changes

Following #673, this PR introduces:

- the possibility for the users to set the Newton solver preconditioner via `F.Settings(preconditioner="...")`
- the possibilty for the users to set a custom Newton solver after the model (`HeatTransferProblem`) initialisation. 

The expected usage of a custom Newton solver is:
```python

my_model.initialise()

class CustomSolver(fenics.NewtonSolver):
     .....

custom_solver = CustomSolver()
custom_solver.parameters["..."] = ...

my_model.h_transport_problem.newton_solver = custom_solver

my_model.run()
``` 
The same can be done for for the `F.ExtrinsicTrap` class:
```python
my_model.traps = [F.ExtrinsicTrap(...), ...]
for trap in my_model.traps:
    if isinstance(trap, F.ExtrinsicTrap)
        trap.newton_solver = CustomSolver()
``` 
and for the `HeatTransferProblem` (but before `my_model.initialise()` for the steady-state case!):

```python
my_model.T = F.HeatTransferProblem(...)
my_model.T.newton_solver = CustomSolver()
``` 

## Types of changes

What types of changes does your code introduce to FESTIM?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring
- [ ] Documentation Update (if none of the other choices apply)
- [x] New tests

## Checklist

- [x] Black formatted
- [x] Unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by KulaginVladimir** [comment]
One thing to mention is that the `info` messages, shown when `log_level=20`, differ with `f.NewtonSolver`, since `f.NonlinearVariationalSolver` [prints](https://bitbucket.org/fenics-project/dolfin/src/1c52e837eb54cc34627f90bde254be4aa8a2ae17/dolfin/fem/NonlinearVariationalSolver.cpp#lines-48) additional messages. A possible workaround is to invoke `f.info` before sovling the problem in FESTIM.

**Comment by RemDelaporteMathurin** [comment]
@KulaginVladimir thanks for this I think it is much needed to give more flexibility.

Would you mind listing the differences - if any - from the user point of view when running the default behaviour?
Are there any differences in absolute errors, usage, etc.

**Comment by RemDelaporteMathurin** [comment]
Also tagging @Allentro as he was interested in this feature

**Comment by KulaginVladimir** [comment]
As I tested, there shouldn't appear significant differences (in errors or usege) with these changes if, for example, `my_model.h_transport_problem.newton_solver` is not provided manually. 

During the model initialisation, a default `fenics.NewtonSolver` object is created and the solver parameters are set from `my_model.settings`: `absolute_tolerance`, `relative_tolerance`, `maximum_iterations`, `linear_solver`, and `preconditioner`, which were kept unchanged (the default preconditioner is set to `default`). Together with the introduced `festim.Problem` class, new changes should reproduce the solbing procedure of `f.NonlinearVariationalProblem` and `f.NonlinearVariationalSolver`, used in the current version of FESTIM. 

As I mentioned above, there is a change in fenics logs (i.e. when `log_level=20`). When `NewtonSolver.solve` is called, it doesn't print "Solving nonlinear variational problem.", whereas `f.NonlinearVariationalSolver` does.

@RemDelaporteMathurin is this what you asked? I also wonder about an example of using a custom solver. Shall we add it somewhere?

File: festim/concentration/traps/traps.py [hunk 1]
```diff
@@ -120,7 +120,12 @@ def solve_extrinsic_traps(self):
                 du_t = f.TrialFunction(trap.density[0].function_space())
                 J_t = f.derivative(trap.form_density, trap.density[0], du_t)
                 problem = festim.Problem(J_t, trap.form_density, [])
+
+                f.begin(
+                    "Solving nonlinear variational problem."
+                )  # Add message to fenics logs
                 trap.newton_solver.solve(problem, trap.density[0].vector())
+                f.end()
```
Comment (KulaginVladimir) [inline festim/concentration/traps/traps.py:128]: So, I added `fenics.begin()`/`fenics.end()` (see [decription](https://fenicsproject.org/olddocs/dolfin/1.4.0/python/programmers-reference/cpp/common/begin.html)) to increase the indentation level of info messages printed by the Newton solver.

**Comment by RemDelaporteMathurin** [comment]
I would like to keep this open while we test it thoroughly. Playing with the solver can be a bit dangerous so I want to make sure we're not merging any breaking stuff.

File: festim/concentration/traps/extrinsic_trap.py [hunk 2]
```diff
@@ -36,18 +38,33 @@ def __init__(
                 If None, the default fenics linear solver will be used ("umfpack").
                 More information can be found at: https://fenicsproject.org/pub/tutorial/html/._ftut1017.html.
                 Defaults to None.
+            preconditioner (str, optional): preconditioning method for the newton solver,
+                options can be veiwed by print(list_krylov_solver_preconditioners()).
```
Comment (RemDelaporteMathurin) [inline festim/concentration/traps/extrinsic_trap.py:42]:
```suggestion
                options can be viewed by print(list_krylov_solver_preconditioners()).
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]
Hi @KulaginVladimir my deepest apologies for taking so much time reviewing this.

I tested it and it seems to work perfectly well!

Several points:
- If users specify a custom solver for say the H transport problem, then the parameters given in `Settings` (absolute, relative tolerances, etc.) will be ignored. This is ok, but we should add a print statement to inform the users
- Can be done in a separate PR but it would be nice to add some documentation and maybe a short demo in the workshop?

File: festim/concentration/traps/extrinsic_trap.py [hunk 3]
```diff
@@ -55,6 +56,23 @@ def __init__(
...
+
+    @newton_solver.setter
+    def newton_solver(self, value):
+        if value is None:
+            self._newton_solver = value
+        elif isinstance(value, NewtonSolver):
+            if self._newton_solver:
+                warnings.warn(
+                    "Settings for the Newton solver will be overwritten", UserWarning
+                )
```
Comment (RemDelaporteMathurin) [inline festim/concentration/traps/extrinsic_trap.py:71]: I wouldn't make this an actual `UserWarning` but rather a print statement. The reason for this is that users could never get rid of this warning. 

Same for other classes.

```suggestion
                print("Settings for the Newton solver will be overwritten")
```

**Review by RemDelaporteMathurin — CHANGES_REQUESTED** [review]

File: festim/concentration/traps/extrinsic_trap.py (hunk 2 above)
Comment (KulaginVladimir) [reply festim/concentration/traps/extrinsic_trap.py:43]: Corrected

File: festim/concentration/traps/extrinsic_trap.py (hunk 3 above)
Comment (KulaginVladimir) [reply festim/concentration/traps/extrinsic_trap.py]: OK

File: festim/nonlinear_problem.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+class Problem(f.NonlinearProblem):
+    """
+    Class to set up a nonlinear variational problem (F(u, v)=0) to solve
+    by the Newton method based on the form of the variational problem, the Jacobian
+    form of the variational problem, and the boundary conditions
+
+    Args:
+        jacobian_form (ufl.Form): the Jacobian form of the variational problem
+        residual_form (ufl.Form): the form of the variational problem
```
Comment (RemDelaporteMathurin) [inline festim/nonlinear_problem.py:12]:
```suggestion
        J (ufl.Form): the Jacobian form of the variational problem
        F (ufl.Form): the form of the variational problem
```

**Review by RemDelaporteMathurin — APPROVED** [review]
Thank you so much for this awesome contribution @KulaginVladimir !!

**Pull request merged**